torch/ao/quantization/fx/prepare.py

```diff
@@ -1165,10 +1175,13 @@ def insert_observers_for_model(
                                     continue
                                 user_node.replace_input_with(node, maybe_output_obs_node)
 
+                            need_remove_observer_ = need_remove_observer(node, modules, node_name_to_target_dtype)
+
                             # for general tensor value ops, we modify the graph
                             # to make all inputs and outputs use the first input's
                             # observer
-                            if is_general_tensor_value_op or is_general_tensor_shape_op or is_reuse_input_qconfig_:
+                            if (is_general_tensor_value_op and need_remove_observer_) or \
+                                    is_general_tensor_shape_op or is_reuse_input_qconfig_:
                                 if not maybe_make_input_output_share_observers(node, model, modules):
```

Comment:
I feel it makes more sense to put `need_remove_observer_` here?

Reply:
we cannot put it here, cuz it would make conflicts on some cases with the flag  'is_reuse_input_qconfig_'

Reply:
I see, this is getting pretty complicated, feels like we need some more refactors and docs..